File: common/contracts/roadmap.py
"""RoadmapContent model — typed artifact for personalized learning roadmaps.

Produced by the RoadmapAgent. Consumed by the renderer's roadmap.eta template.
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from common.contracts.components import ContentComponent, StatCard


class NavItem(BaseModel):
    label: str
    href: str
    group: str = "a"


class LegendItem(BaseModel):
    color: str
    label: str


class RoadmapHero(BaseModel):
    eyebrow: str = ""
    title: str
    lede: str = ""
    stamp: str = ""
    stats: list[StatCard] = Field(default_factory=list)


class RoadmapSidebar(BaseModel):
    title: str
    subtitle: str
    stats: list[StatCard] = Field(default_factory=list)
    nav: list[NavItem] = Field(default_factory=list)
    legend: list[LegendItem] = Field(default_factory=list)


class RoadmapSection(BaseModel):
    id: str
    title: str
    subtitle: str | None = None
    tag_num: str | None = None
    components: list[ContentComponent] = Field(default_factory=list)


class RoadmapContent(BaseModel):
    artifact_type: Literal["roadmap"] = "roadmap"
    title: str
    theme: str = "default"
    hero: RoadmapHero
    sections: list[RoadmapSection] = Field(default_factory=list)
    sidebar: RoadmapSidebar
    accessibility: dict = Field(default_factory=lambda: {"language": "vi"})

File: packages/renderer/src/contracts/roadmap.ts
/**
 * Roadmap artifact data contract.
 *
 * Mirrors RoadmapContent Python Pydantic model (common/contracts/roadmap.py).
 * Used when artifact_type == "roadmap".
 */

import type { ContentComponent } from "./components.js";

export interface StatCard {
  label: string;
  value: string;
  variant?: "target" | "now" | "default";
}

export interface NavItem {
  label: string;
  href: string;
  group?: string;
}

export interface LegendItem {
  color: string;
  label: string;
}

export interface RoadmapSidebar {
  title: string;
  subtitle: string;
  stats?: StatCard[];
  nav?: NavItem[];
  legend?: LegendItem[];
}

export interface RoadmapHero {
  eyebrow?: string;
  title: string;
  lede?: string;
  stamp?: string;
  stats?: StatCard[];
}

export interface RoadmapSection {
  id: string;
  title: string;
  subtitle?: string;
  tag_num?: string;
  components?: ContentComponent[];
}

export interface RoadmapData {
  title: string;
  theme?: string;
  hero: RoadmapHero;
  sections?: RoadmapSection[];
  sidebar: RoadmapSidebar;
  accessibility?: { language?: string };
}
